core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java

```diff
@@ -167,25 +142,109 @@ public PublishVideoOperation(InputSocket.Factory inputSocketFactory) {
+    copyJavaCvToOpenCvMat(input, publishMat);
+    // Make sure the output resolution is up to date. Might not be needed, depends on cscore updates
+    serverSource.setResolution(input.size().width(), input.size().height());
+    serverSource.putFrame(publishMat);
+    if (lastFrame != -1) {
+      long dt = now - lastFrame;
+      serverSource.setFPS((int) (1e9 / dt));
     }
+    lastFrame = now;
   }
 
   @Override
   public synchronized void cleanUp() {
-    // Stop the video server if there are no Publish Video steps left
-    serverThread.interrupt();
     numSteps--;
+    if (cscoreLoaded) {
+      availablePorts.addFirst(server.getPort());
+      ourTable.getKeys().forEach(ourTable::delete);
+      serverSource.setConnected(false);
+      serverSource.free();
+      server.free();
+    }
+  }
+
+  /**
+   * Copies the data from a JavaCV Mat wrapper object into an OpenCV Mat wrapper object so it's
+   * usable by the {@link CvSource} for this operation.
+   *
+   * <p>Since the JavaCV and OpenCV bindings both target the same native version of OpenCV, this is
+   * implemented by simply changing the OpenCV Mat's native pointer to be the same as the one for
+   * the JavaCV Mat. This prevents memory copies and resizing/reallocating memory for the OpenCV
+   * wrapper to fit the source image. Updating the pointer is a simple field write (albeit via
+   * reflection), which is much faster and easier than allocating and copying byte buffers.</p>
+   *
+   * <p>A caveat to this approach is that the memory layout used by the OpenCV binaries bundled with
+   * both wrapper libraries <i>must</i> be identical. Using the same OpenCV version for both
+   * libraries should be enough.</p>
+   *
+   * @param javaCvMat the JavaCV Mat wrapper object to copy from
+   * @param openCvMat the OpenCV Mat wrapper object to copy into
+   * @throws RuntimeException if the OpenCV native pointer could not be set
+   */
+  private static void copyJavaCvToOpenCvMat(opencv_core.Mat javaCvMat, Mat openCvMat)
+      throws RuntimeException {
+    // Make the OpenCV Mat object point to the same block of memory as the JavaCV object.
+    // This requires no data transfers or copies and is O(1) instead of O(n)
+    if (javaCvMat.address() != openCvMat.nativeObj) {
+      try {
+        Field nativeObjField = Mat.class.getField("nativeObj");
+        nativeObjField.setAccessible(true);
+        nativeObjField.setLong(openCvMat, javaCvMat.address());
+      } catch (ReflectiveOperationException e) {
+        logger.log(Level.WARNING, "Could not set native object pointer", e);
+        throw new RuntimeException("Could not copy the image", e);
+      }
+    }
   }
+
+  /**
+   * Multi platform method for getting the hostname of the local computer. cscore's
+   * {@link CameraServerJNI#getHostname() getHostName() function} only works on Linux, so we need to
+   * implement the method for Windows and Mac ourselves.
+   */
+  private static String getHostName() {
```

Comment:
Replace with: `InetAddress.getLocalHost().getHostName()`???

Reply:
Oooh. That's much better 👍